### ISSUE TITLE
synchronization is not working (fixed on v1.1.6)

### ISSUE THREAD

**Issue by insinfo** [gmpassos/shared_map#11]
build this example of a very simple rest API with two endpoints, one '/' that returns ok and the other that displays the '/metrics' counter, it turns out that if I make a simultaneous access with the command `xargs -I % -P 8 curl "http://192.168.66.123:3161/" < <(printf '%s\n' {1..4})` this counter was supposed to have a value of 400 but instead it gives me different values ​​for each execution

```dart
import 'dart:convert';
import 'dart:io';
import 'dart:isolate';
import 'package:shared_map/shared_map.dart';
//run with: xargs -I % -P 8 curl "http://192.168.66.123:3161/" < <(printf '%s\n' {1..4})

void main(List<String> args) async {
  final store1 = SharedStore('t1');
  final counter = await store1.getSharedMap<int, int>('m1');
  final counterReference = counter!.sharedReference();
  counter.put(0, 1);

  final args = ['0.0.0.0', 3161];
  final concurrency = 2;
  for (var i = 0; i < concurrency - 1; i++) {
    await Isolate.spawn(startIsolate, [i, ...args, counterReference]);
  }
  startIsolate([concurrency - 1, ...args, counterReference]);
}

void startIsolate(List args) async {
  final id = args[0] as int;
  final address = args[1] as String;
  final port = args[2] as int;
  final counterReference = args[3] as SharedMapReference;
  final server = await HttpServer.bind(address, port, shared: true);
  server.listen((req) {
    log(req, id);
    handler(req, req.response, counterReference);
  });
  print('server $id ${server.address.address}:${server.port}');
}

void handler(
    HttpRequest req, HttpResponse res, SharedMapReference counterR) async {
  final counter = SharedMap<int, int>.fromSharedReference(counterR);

  if (req.uri.path != '/metrics') {
    // increment
    counter.put(0, (await counter.get(0))! + 1);
  }
  switch (req.uri.path) {
    case '/':
      res.withString('ok');
      break;
    case '/metrics':
      res.withString('count ${(await counter.get(0))!}');
      break;
    default:
      res.notFound();
  }
}

void log(HttpRequest req, int id) {
  print('$id ${req.method} ${req.uri.path}');
}

extension HttpResponseExtension on HttpResponse {
  void withJson(Object? object) {
    this.statusCode = 200;
    this.writeln(jsonEncode(object));
    this.close();
  }

  void withString(String str) {
    this.statusCode = 200;
    this.writeln(str);
    this.close();
  }

  void ok() {
    this.statusCode = 200;
    this.close();
  }

  void notFound() {
    this.statusCode = 404;
    this.close();
  }
}

```

**Comment by insinfo** [comment]
I do it with a pointer and the native_synchronization package works correctly

```dart
import 'dart:convert';
import 'dart:ffi';
import 'dart:io';
import 'dart:isolate';

import 'package:ffi/ffi.dart';
import 'package:native_synchronization/primitives.dart';
import 'package:native_synchronization/sendable.dart';

//run with: xargs -I % -P 8 curl "http://192.168.66.123:3161/" < <(printf '%s\n' {1..4})

void main(List<String> args) async {
  final counter = calloc<Uint64>();
  final mutex = Mutex();

  final args = ['0.0.0.0', 3161];
  final concurrency = 2;
  for (var i = 0; i < concurrency - 1; i++) {
    await Isolate.spawn(
        startIsolate, [i, ...args, counter.address, mutex.asSendable]);
  }
  startIsolate([concurrency - 1, ...args, counter.address, mutex.asSendable]);
}

void startIsolate(List args) async {
  final id = args[0] as int;
  final address = args[1] as String;
  final port = args[2] as int;
  final counterAddress = args[3] as int;
  final sendableMutex = args[4] as Sendable<Mutex>;

  final server = await HttpServer.bind(address, port, shared: true);
  server.listen((req) {
    log(req, id);
    handler(req, req.response, counterAddress, sendableMutex);
  });
  print('server $id ${server.address.address}:${server.port}');
}

void handler(HttpRequest req, HttpResponse res, int counterAddress,
    Sendable<Mutex> sendableMutex) async {
  final counter = Pointer<Uint64>.fromAddress(counterAddress);
  final mutex = sendableMutex.materialize();

  if (req.uri.path != '/metrics') {
    // increment
    mutex.runLocked(() {
      counter.value = counter.value + 1;
    });
  }

  switch (req.uri.path) {
    case '/':
      res.withString('ok');
      break;
    case '/metrics':
      res.withString('count ${counter.value}');
      break;
    default:
      res.notFound();
  }
}

void log(HttpRequest req, int id) {
  print('$id ${req.method} ${req.uri.path}');
}

extension HttpResponseExtension on HttpResponse {
  void withJson(Object? object) {
    this.statusCode = 200;
    this.writeln(jsonEncode(object));
    this.close();
  }

  void withString(String str) {
    this.statusCode = 200;
    this.writeln(str);
    this.close();
  }

  void ok() {
    this.statusCode = 200;
    this.close();
  }

  void notFound() {
    this.statusCode = 404;
    this.close();
  }
}

```

**Comment by gmpassos** [comment]
I have made some adjustment on your original example:

```dart
import 'dart:convert';
import 'dart:io';
import 'dart:isolate';

import 'package:shared_map/shared_map.dart';
//run with: xargs -I % -P 8 curl "http://192.168.66.123:3161/" < <(printf '%s\n' {1..4})

void main(List<String> args) async {
  final store1 = SharedStore('t1');
  final counter = await store1.getSharedMap<int, int>('m1');
  final counterReference = counter!.sharedReference();
  counter.put(0, 1);

  final args = ['0.0.0.0', 3161];
  final concurrency = 2;

  for (var i = 0; i < concurrency - 1; i++) {
    await Isolate.spawn(startIsolate, [i, ...args, counterReference]);
  }

  startIsolate([concurrency - 1, ...args, counterReference]);
}

void startIsolate(List args) async {
  final id = args[0] as int;
  final address = args[1] as String;
  final port = args[2] as int;
  final counterReference = args[3] as SharedMapReference;
  final server = await HttpServer.bind(address, port, shared: true);

  final counter = SharedMap<int, int>.fromSharedReference(counterReference);

  server.listen((req) {
    log(req, id);
    handler(req, req.response, counter);
  });

  print('server $id ${server.address.address}:${server.port}');
}

void handler(
    HttpRequest req, HttpResponse res, SharedMap<int, int> counter) async {
  final path = req.uri.path;

  var count = await counter.get(0);

  switch (path) {
    case '/':
      var count2 = await counter.put(0, count! + 1);
      res.withString('ok (count: $count -> $count2)');
      break;
    case '/metrics':
      res.withString('count $count');
      break;
    default:
      res.notFound();
  }
}

void log(HttpRequest req, int id) {
  print('$id ${req.method} ${req.uri.path}');
}

extension HttpResponseExtension on HttpResponse {
  void withJson(Object? object) {
    statusCode = 200;
    writeln(jsonEncode(object));
    close();
  }

  void withString(String str) {
    statusCode = 200;
    writeln(str);
    close();
  }

  void ok() {
    statusCode = 200;
    close();
  }

  void notFound() {
    statusCode = 404;
    close();
  }
}

```

Request outputs:
```text
 xargs -I % -P 8 curl "http://127.0.0.1:3161/" < <(printf '%s\n' {1..4})
ok (count: 1 -> 2)
ok (count: 2 -> 3)
ok (count: 3 -> 4)
ok (count: 3 -> 4)
```

Since the "increment" computation is not made inside the main `Isolate`, there's no "synchronization", this is why it's possible to get duplicated/missed counts.

**Comment by gmpassos** [comment]
Take a look into this branch:

https://github.com/gmpassos/shared_map/pull/12

Best regards.

**Comment by gmpassos** [comment]
See the new version, with the method `SharedMap.update`:

`shared_map: ^1.1.6`

**Comment by gmpassos** [comment]
Here's your code using the new `update` method (`shared_map: ^1.1.6`):

```dart
import 'dart:convert';
import 'dart:io';
import 'dart:isolate';

import 'package:shared_map/shared_map.dart';
//run with: xargs -I % -P 8 curl "http://192.168.66.123:3161/" < <(printf '%s\n' {1..4})

void main(List<String> args) async {
  final store1 = SharedStore('t1');
  final counter = await store1.getSharedMap<int, int>('m1');
  final counterReference = counter!.sharedReference();
  counter.put(0, 1);

  final args = ['0.0.0.0', 3161];
  final concurrency = 2;

  for (var i = 0; i < concurrency - 1; i++) {
    await Isolate.spawn(startIsolate, [i, ...args, counterReference]);
  }

  startIsolate([concurrency - 1, ...args, counterReference]);
}

void startIsolate(List args) async {
  final id = args[0] as int;
  final address = args[1] as String;
  final port = args[2] as int;
  final counterReference = args[3] as SharedMapReference;
  final server = await HttpServer.bind(address, port, shared: true);

  final counter = SharedMap<int, int>.fromSharedReference(counterReference);

  server.listen((req) {
    log(req, id);
    handler(req, req.response, counter);
  });

  print('server $id ${server.address.address}:${server.port}');
}

void handler(
    HttpRequest req, HttpResponse res, SharedMap<int, int> counter) async {
  final path = req.uri.path;

  var count = await counter.get(0);

  switch (path) {
    case '/':
      // Un-synchronized computation:
      //var count2 = await counter.put(0, count! + 1);

      // Synchronized computation:
      var count2 = await counter.update(0, _increment);

      res.withString('ok (count: $count -> $count2)');
      break;
    case '/metrics':
      res.withString('count $count');
      break;
    default:
      res.notFound();
  }
}

int _increment(int k, int? v) => (v ?? 0) + 1;

void log(HttpRequest req, int id) {
  print('$id ${req.method} ${req.uri.path}');
}

extension HttpResponseExtension on HttpResponse {
  void withJson(Object? object) {
    statusCode = 200;
    writeln(jsonEncode(object));
    close();
  }

  void withString(String str) {
    statusCode = 200;
    writeln(str);
    close();
  }

  void ok() {
    statusCode = 200;
    close();
  }

  void notFound() {
    statusCode = 404;
    close();
  }
}

```

**Comment by insinfo** [comment]
With the use of the update method it is now working correctly, what synchronization logic are you using? I didn't look at your source code but I saw that you are not using the native_synchronization package so I think you are not using Mutex right?

```console
xargs -I % -P 8 curl "http://192.168.18.31:3161/" < <(printf '%s\n' {1..4000})
...
ok (count: 3993 -> 3994)
ok (count: 3994 -> 3995)
ok (count: 3995 -> 3996)
ok (count: 3996 -> 3997)
ok (count: 3997 -> 3998)
ok (count: 3998 -> 3999)
ok (count: 3999 -> 4000)
```

**Comment by gmpassos** [comment]
```
what synchronization logic are you using?
...
I saw that you are not using the native_synchronization package 
```

I believe this is the most crucial point. When attempting to establish parallelism without multi-threading memory manipulation, we can't apply the same paradigms/techniques of data manipulation and synchronization. Therefore, using a "native" mutex in the Dart SDK (similar to the JDK) is not the ideal approach.

What we need is not a **MUTEX** but a "critical section" or a "safe" region for manipulating the data. Therefore, what I've done is to ensure that the `updater` lambda/anonymous/closure function you pass to the `SharedMap.update` method is executed in a safe region, in this case, within the main `SharedMap` instance (main/original Isolate).

In other words, the `updater` lambda/anonymous/closure function is passed as a parameter to the main Isolate (supported since Dart 2.15). Subsequently, the `updater` code is executed in the main `Isolate` in a thread-safe manner.

**Comment by gmpassos** [comment]
You can read more about in this Medium article:

https://medium.com/@gracilianomp/using-shared-map-for-concurrency-in-dart-d15eafcb715a
